File: year2012/CEOI/day1/jobs.pas
const maxn=100020; maxm=1000000; inf=maxlongint;
type rec=record v,nxt:longint; end;
     save=record id,day:longint; end;
var n,d,m,i,j,k,x,le,ri,mid,ans,tot,top,cnt,last:longint;
    edge:array[0..maxn]of longint;
    heap:array[0..maxm]of longint;
    a:array[0..maxm]of longint;
    g:array[0..maxm]of rec;
    fans,sta:array[0..maxm]of save;
procedure addedge(x,y:longint);
begin
inc(tot); g[tot].v:=y; g[tot].nxt:=edge[x]; edge[x]:=tot;
end;
procedure up(x:longint);
var swap:longint;
begin
while x>0 do
  begin
  if a[heap[x]]<a[heap[x div 2]] then
     begin
     swap:=heap[x]; heap[x]:=heap[x div 2]; heap[x div 2]:=swap;
     x:=x div 2;
     end
  else break;
  end;
end;
procedure down(x:longint);
var numa,numb,swap,opt:longint;
begin
while x*2<=top do
  begin
  numa:=a[heap[x*2]]; if x*2+1<=top then numb:=a[heap[x*2+1]] else numb:=inf;
  if numa<numb then opt:=0 else opt:=1;
  if a[heap[x]]>a[heap[x*2+opt]] then
     begin
     swap:=heap[x]; heap[x]:=heap[x*2+opt]; heap[x*2+opt]:=swap;
     x:=x*2+opt;
     end
  else break;
  end;
end;
procedure insert(nowid:longint);
begin
inc(top); heap[top]:=nowid;
up(1);
end;
function check(lim:longint):boolean;
var i,j,p:longint;
begin
top:=0; cnt:=0;
for i:=1 to n do
    begin
    if (top>0)and(a[heap[1]]<i-d) then exit(false);
    p:=edge[i];
    while p<>0 do
      begin
      insert(g[p].v);
      p:=g[p].nxt;
      end;
    j:=0;
    while (top>0)and(j<lim) do
      begin
      inc(cnt); inc(j);
      sta[cnt].id:=heap[1]; sta[cnt].day:=i;
      heap[1]:=heap[top]; heap[top]:=0; dec(top);
      down(1);
      end;
    end;
exit(true);
end;
begin
assign(input,'jobs.in');
reset(input);
assign(output,'jobs.out');
rewrite(output);
read(n,d,m);
for i:=1 to m do
    begin
    read(x);
    addedge(x,i);
    a[i]:=x;
    end;
le:=1; ri:=m; ans:=-1;
while le<=ri do
  begin
  mid:=(le+ri) div 2;
  if check(mid) then begin ans:=mid; fans:=sta; ri:=mid-1; end
     else le:=mid+1;
  end;
writeln(ans);
last:=0;
i:=1;
while i<=m do
  begin
  j:=i;
  while (j+1<=m)and(fans[j+1].day=fans[i].day) do inc(j);
  for k:=last+1 to fans[i].day-1 do writeln(0);
  for k:=i to j do write(fans[k].id,' ');
  write(0);
  writeln;
  last:=fans[i].day;
  i:=j+1;
  end;
for i:=last+1 to n do writeln(0);
close(input);
close(output);
end.

File: year2012/CEOI/day1/jobs.cpp
#include<cstdio>
#include<iostream>
#define maxn 111111
#define maxm 1111111
using namespace std;
int fst[maxn],nxt[maxm];
int t,n,d,m,k,l,r,mid;
int tot[maxn];
bool pan(int k)
{
    int m=0,cnt=0;
    for (int i=1;i<=n-d;i++)
    {
        if (cnt+(m>0)<i) cnt=i-1,m=0;
        cnt+=(tot[i]+m)/k;
        m=(tot[i]+m)%k;
        if (cnt+(m>0)-i>d) return 0;
    }
    return 1;
}
int main()
{
    freopen("jobs.in","r",stdin);
    freopen("jobs.out","w",stdout);
    scanf("%d%d%d",&n,&d,&m);
    for (int i=1;i<=m;i++)
    {
        scanf("%d",&t);
        tot[t]++;
        nxt[i]=fst[t];
        fst[t]=i;
    }
    l=0,r=m;
    while (l<r-1)
    {
        mid=(l+r)>>1;
        if (pan(mid)) r=mid;else l=mid;   
    }
    printf("%d\n",r);
    t=1;
    while (!k) k=fst[t++];
    for (int i=1;i<=n;i++)
    {
        for (int j=1;j<=r;j++)
        {
            if (k && t-1<=i) printf("%d ",k);
            else break;
            if (nxt[k]) k=nxt[k];
            else 
            { 
                k=fst[t++];
                while (!k && t<=n) k=fst[t++];
            }
        }
        printf("0\n");
    }       
    return 0;   
}
